core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/DiskBasedChunkCache.java

```diff
@@ -0,0 +1,133 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.chunkmanager.cache;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+
+import io.aiven.kafka.tieredstorage.chunkmanager.ChunkKey;
+import io.aiven.kafka.tieredstorage.chunkmanager.ChunkManager;
+
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.Weigher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+
+public class DiskBasedChunkCache extends ChunkCache<Path> {
+
+    private static final Logger log = LoggerFactory.getLogger(DiskBasedChunkCache.class);
+    static final String TEMP_CACHE_DIRECTORY = "temp";
+    static final String CACHE_DIRECTORY = "cache";
+
+    private DiskBasedChunkCacheConfig config;
+
+    public DiskBasedChunkCache(final ChunkManager chunkManager) {
+        super(chunkManager);
+    }
+
+    @Override
+    public InputStream cachedChunkToInputStream(final Path cachedChunk) {
+        try {
+            return Files.newInputStream(cachedChunk);
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Writes the chunk to specified directory on disk.
+     * To be able to understand if the files are fully written to disk during a potential recovery process,
+     * this implementation initially writes them to temporary directory and then atomically moves them
+     * to the actual caching directory.
+     */
+    @Override
+    public Path cacheChunk(final ChunkKey chunkKey, final InputStream chunk) throws IOException {
+        final Path tempChunkPath = config.baseCachePath()
+                .resolve(TEMP_CACHE_DIRECTORY)
+                .resolve(chunkKey.uuid + "-" + chunkKey.chunkId);
+        if (Files.notExists(tempChunkPath.getParent())) {
+            Files.createDirectories(tempChunkPath.getParent());
+        }
+        final Path tempCached = writeToDisk(chunk, tempChunkPath);
+        log.debug("Chunk file has been stored to temporary caching directory {}", tempCached);
+        final Path cachedChunkPath = config.baseCachePath()
+                .resolve(CACHE_DIRECTORY)
+                .resolve(chunkKey.uuid + "-" + chunkKey.chunkId);
+        if (Files.notExists(cachedChunkPath.getParent())) {
+            Files.createDirectories(cachedChunkPath.getParent());
+        }
+        try {
+            final Path newPath = Files.move(tempCached, cachedChunkPath, ATOMIC_MOVE);
+            log.debug("Chunk file has been moved to cache directory {}", newPath);
+            return newPath;
+        } finally {
+            // Cleaning up the temp directory in case of exceptions during move.
+            Files.deleteIfExists(tempCached);
```

Comment:
Maybe some error logging here would be good, too

Reply:
Added